[Security Solution] - use expandable flyout in timeline

PhilippeOberti opened the pull request:
## Summary

This PR enables Timeline to have its own expandable flyout, independent from the Security Solution app level flyout. This previous [PR](https://github.com/elastic/kibana/pull/176457) added support to the `kbn-expandable-flyout` package to be able to show multiple flyout at the same time.

This PR focuses on:
- changing the location where we use the `<ExpandableFlyoutProvider>` to wrap only the Security Solution pages. instead of the entire Security Solution application
- wrapping `Timeline` with its own `<ExpandableFlyoutProvider>`
- opening the new expandable flyout from timeline for alerts, users and host!!

### Notes

- the new expandable flyout opened from `Timeline` is saved in the url under a new `timelineFlyout` parameter. The other Security Solution flyout stays saved under `eventFlyout` so both can be displayed at the same time
- introduced a new feature flag called `expandableTimelineFlyoutEnabled`, enabled to `false` by default. The code uses this flag in combination with the already existing `securitySolution:enableExpandableFlyout` advanced settings to toggle on/off the expandable flyout in `Timeline`

I had to make a small change to the timeline `z-index` value, from its hardcoded value of `1000` to now `1001`. At the same time I'm setting the `z-index` of the `eventFlyout` to `1000` and the `z-index` of the `timelineFlyout` to `1002`. Finally, I changed the `z-index` of all the other flyouts opened from the `Take action` button in the footer (like _Add to new case_, _Add rule exception_...) to `1003`. This was the only way to get the order correct after page refresh, and also work in Serverless which has a different top level layout (for the top bar and the left navigation panel).

### Testing

Make sure you have the flag turned on in your `kibana.yml` file: `xpack.securitySolution.enableExperimental: ['expandableTimelineFlyoutEnabled']`

While the code changes are pretty minimal, emphasize should be done on testing:
- verify that the non-timeline expandable flyout behavior hasn't changed (in the Explore, Rule Preview, Alerts pages...)
- verify the advanced settings correctly toggles on and off the expandable flyout for both the Security Solution app and Timeline
- verify the expandable flyout correctly shows up in Timeline on all the pages where Timeline is present
- verify both flyout can be shown at the same time
- verify refreshing the page reloads correctly all flyouts open
- verify copying the url or using the Share Alert button at the top can be correctly reopened in another tab

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

### TODO

- [x] fix unit tests
- [x] fix e2e tests
- [x] figure out how to fix the UI in serverless (the flyout is currently behind the left navigation)

Before change:

https://github.com/elastic/kibana/assets/17276605/3b89f87f-d1fa-4635-8ff7-855795eb3796

After change:

https://github.com/elastic/kibana/assets/17276605/6b341f11-054c-4885-b496-006f37b8d572

Serverless

https://github.com/elastic/kibana/assets/17276605/df414140-31df-4941-869e-c177cfedd805

https://github.com/elastic/security-team/issues/7464

Details of file changed in [the comment below](https://github.com/elastic/kibana/pull/177087#issuecomment-1988642537)

PhilippeOberti committed:
[Security Solution] separate main expandable flyout from timeline expandable flyout

elasticmachine commented:
Pinging @elastic/security-threat-hunting-investigations (Team:Threat Hunting:Investigations)

PhilippeOberti commented:
\ci

christineweng commented on x-pack/plugins/security_solution/common/experimental_features.ts:
```diff
@@ -92,6 +92,10 @@ export const allowedExperimentalValues = Object.freeze({
+  /**
+   * Enables expandable flyout in timeline
+   */
+  expandableTimelineFlyoutEnabled: true,
```
Is there a way to disable the flag outside of this file? To me it's a little risky to default true, since once merged it will take affect in serverless

PhilippeOberti replied:
good point, I can switch it back to `false` by default!

christineweng commented on x-pack/plugins/security_solution/public/flyout/index.tsx:
```diff
@@ -90,8 +90,30 @@ const expandableFlyoutDocumentsPanels: ExpandableFlyoutProps['registeredPanels']
+  <ExpandableFlyout
+    registeredPanels={expandableFlyoutDocumentsPanels}
+    paddingSize="none"
+    customStyles={{ 'z-index': 1001 }}
```
The timeline flyout will cover flyouts such as `Add to new case`, `Add Endpoint exception` and `Add rule exception` from take action. The new case flyout will stay after closing the timeline flyout. But endpoint and rule exception are not usable, because they always close automatically when timeline flyout is closed. 

https://github.com/elastic/kibana/assets/18648970/f1926861-e119-44be-82c7-cb73f529aa75

PhilippeOberti replied:
crap, I thoughts had looked at some of these and their `z-index` was higher (I remember seeing on at like `8000` lol)...
Thanks for testing this, I'll look into it!

PhilippeOberti replied:
very interesting. I had tested this for the `eventFlyout` and it was working correctly. I saw that those have a custom `z-index` (set to `8000` for the Create case flyout for example) so I did not worry about it.
I'm not sure why it's not working correctly for the `timelineFlyout` as the `z-index: 8000` is there...
![Screenshot 2024-02-21 at 5 10 37 PM](https://github.com/elastic/kibana/assets/17276605/920c9cfe-a396-4740-90b3-9068fd675227)

PhilippeOberti replied:
ok after more investigation, the reason this is happening is because all of these flyouts opened from the `Take action` (Create new case, Add Endpoint exception, Add rule exception...) are using an `overlayMask` from EUI. This which adds a class to the very top parent of those flyout which sets a `z-index` to 1000. Internally they set a `z-index` to 8000 to the actual flyout component, but that doesn't matter as a parent has another `z-index` set. I was confused at first by that.

I decided to - instead of making changes to all of these flyouts that we don't own and would be used in other places we're not aware of - make a small change to the `z-index` values we can control:
- the `eventFlyout` `z-index` value is now `997` instead of `999`
- the `timeline` `z-index` value is now `998` instead of `1000`
- the `timelineFlyout` `z-index` value is now `999` instead of `1000`

This should guarantee that these are still playing nice with each other, while staying behind the other `Take action` flyouts or other potential modals being open...

PhilippeOberti committed:
PR comments:
- turn feature flag to off
- enable user and host details expandable flyout
- fix z-index issue with take action flyouts

machadoum commented on config/kibana.yml:
```diff
@@ -181,3 +181,5 @@
 # This value must be a whole number greater than zero. Defaults to 100_000
 #unifiedSearch.autocomplete.valueSuggestions.terminateAfter: 100000
+
+#xpack.securitySolution.enableExperimental: ['expandableTimelineFlyoutEnabled']
```
Is this change intentional? It is the first time I have seen a reference to `enableExperimental ` inside this file.

PhilippeOberti replied:
no it's not, good catch! 😆

machadoum reviewed (approved):
The Entity flyout code looks good to me.

I did some exploratory tests and found only one issue. I am not blocking the PR because the feature is behind an experimental flag, and we can fix it later. Could you open a bug 
 ticket to address it?

Issue: The entity flyout opened inside the timeline renders cell actions with the wrong context. So, the filter action adds a filter to the background page instead of adding it to the timeline.
![Feb-23-2024 12-45-18](https://github.com/elastic/kibana/assets/1490444/c814ed39-0f60-4185-8c3c-0c3b1873d895)

PhilippeOberti committed:
remove kibana.yml changes
remove unnecessary copy logic

christineweng commented on x-pack/plugins/security_solution/public/flyout/document_details/right/components/header_actions.tsx:
```diff
@@ -82,8 +76,8 @@ export const HeaderActions: VFC = memo(() => {
                   data-test-subj={SHARE_BUTTON_TEST_ID}
-                  onClick={() => copyFunction(copy, alertDetailsLink, modifier)}
-                  onKeyDown={() => copyFunction(copy, alertDetailsLink, modifier)}
+                  onClick={() => copyFunction(copy, alertDetailsLink)}
```
nit: if url modification is no longer necessary, we probably don't need the [helper function](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/public/flyout/shared/utils/copy_to_clipboard.tsx#L18) and just use eui copy?

christineweng reviewed (approved):
LGTM 🚀

PhilippeOberti committed:
remove unused copy_to_clipboard files

ashokaditya reviewed (commented):
Thanks for the changes! I tested it out and the flyouts work as expected. I've a few suggestions for improvements.

PhilippeOberti commented on x-pack/plugins/security_solution/public/flyout/index.tsx:
```diff
@@ -90,8 +90,32 @@ const expandableFlyoutDocumentsPanels: ExpandableFlyoutProps['registeredPanels']
 ));
 
 SecuritySolutionFlyout.displayName = 'SecuritySolutionFlyout';
+
+/**
+ * Flyout used in Timeline
+ * We set the z-index to 999 to ensure it is always rendered above Timeline (which has a z-index of 998)
+ * and also behind other flyouts like Create new Case, Add Endpoint exception, Add rule exception, etc...
+ * opened from the Take action button in the footer. These latter flyouts use an overlay mask and that index is set to 1000 so we need to be below that.
+ */
```
great idea! 👍

PhilippeOberti replied:
I did something slightly different, instead of having those numbers as constants, I'm getting the default flyout `z-index` (`1000`) from EUI, then adding 1, 2 or 3. I did not extract this into a function because I didn't really know where to put it (the code impacted lives in different plugins. Instead I added comments everywhere to explain the decision.
@ashokaditya let me know if that is ok with you!

PhilippeOberti committed:
fix layout issue in serverless

PhilippeOberti commented on x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/host_name.tsx:
```diff
@@ -65,49 +69,68 @@ const HostNameComponent: React.FC<Props> = ({
+        }
+      };
+
+      if (isTimelineScope(timelineID)) {
```
done!

e40pud commented on x-pack/plugins/security_solution/public/detection_engine/rule_exceptions/components/add_exception_flyout/index.tsx:
```diff
@@ -493,7 +495,13 @@ export const AddExceptionFlyout = memo(function AddExceptionFlyout({
+      onClose={handleCloseFlyout}
+      data-test-subj="addExceptionFlyout"
+      // EUI TODO: This z-index override of EuiOverlayMask is a workaround, and ideally should be resolved with a cleaner UI/UX flow long-term
+      maskProps={{ style: `z-index: ${(euiTheme.levels.flyout as number) + 3}` }} // we need this flyout to be above the timeline flyout (which has a z-index of 1002)
```
Maybe we could put all `(euiTheme.levels.flyout as number) + #` in a separate place. Maybe be hook which returns all the values or returns a specific one based on passed parameter: `ExceptionFlyout`, `TimelineFlyout` etc...

I'm afraid that it will be hard to track z-index in different places and understand what is an appropriate value if we add a flyout somewhere else.

PhilippeOberti replied:
yeah, you're the second person mentioning this. The challenge is this would be used across plugins (cases and security_solution) and I don't really want to create a package just for that... Any idea where would be a good place for this?

e40pud reviewed (approved):
Tested alert details flyout locally - everything works. DE changes LGTM! Left one comment for potential improvement.

ashokaditya reviewed (approved):
Thanks for the changes again @PhilippeOberti. This is looking good to 🚢

PhilippeOberti commented:
## Files by Code Owner

### elastic/response-ops

* x-pack/plugins/cases/public/components/create/flyout/create_case_flyout.tsx

### elastic/security-defend-workflows

* x-pack/plugins/security_solution/public/management/components/page_overlay/page_overlay.tsx
* x-pack/plugins/security_solution/public/management/cypress/e2e/response_actions/alerts_response_console.cy.ts
* x-pack/plugins/security_solution/public/management/cypress/screens/alerts.ts

### elastic/security-detection-engine

* x-pack/plugins/security_solution/public/detection_engine/rule_exceptions/components/add_exception_flyout/index.tsx

### elastic/security-engineering-productivity

* x-pack/test/security_solution_cypress/cypress/e2e/investigations/timelines/toggle_column.cy.ts

### elastic/security-solution

* x-pack/plugins/security_solution/common/experimental_features.ts
* x-pack/plugins/security_solution/public/app/home/template_wrapper/index.tsx
* x-pack/plugins/security_solution/public/common/hooks/use_url_state.ts
* x-pack/plugins/security_solution/public/detection_engine/rule_exceptions/components/add_exception_flyout/index.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/right/components/header_actions.test.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/right/components/header_actions.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/right/tabs/json_tab.test.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/right/tabs/json_tab.tsx
* x-pack/plugins/security_solution/public/flyout/index.tsx
* x-pack/plugins/security_solution/public/flyout/shared/utils/copy_to_clipboard.test.tsx
* x-pack/plugins/security_solution/public/flyout/shared/utils/copy_to_clipboard.tsx
* x-pack/plugins/security_solution/public/management/components/page_overlay/page_overlay.tsx
* x-pack/plugins/security_solution/public/management/cypress/e2e/response_actions/alerts_response_console.cy.ts
* x-pack/plugins/security_solution/public/management/cypress/screens/alerts.ts
* x-pack/plugins/security_solution/public/timelines/components/modal/index.styles.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/events/stateful_event.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/host_name.test.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/host_name.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/user_name.test.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/user_name.tsx

### elastic/security-threat-hunting-investigations

* packages/kbn-expandable-flyout/src/index.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/right/components/header_actions.test.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/right/components/header_actions.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/right/tabs/json_tab.test.tsx
* x-pack/plugins/security_solution/public/flyout/document_details/right/tabs/json_tab.tsx
* x-pack/plugins/security_solution/public/flyout/shared/utils/copy_to_clipboard.test.tsx
* x-pack/plugins/security_solution/public/flyout/shared/utils/copy_to_clipboard.tsx
* x-pack/plugins/security_solution/public/timelines/components/modal/index.styles.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/events/stateful_event.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/host_name.test.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/host_name.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/user_name.test.tsx
* x-pack/plugins/security_solution/public/timelines/components/timeline/body/renderers/user_name.tsx
* x-pack/test/security_solution_cypress/cypress/e2e/investigations/timelines/toggle_column.cy.ts

cnasikas reviewed (approved):
ResponseOps LGTM!

PhilippeOberti committed:
update z-index for endpoint response and browser field flyouts

mattkime reviewed (approved):
codeowner changes lgtm

PhilippeOberti committed:
remove unnecessary case flyout z-index
